patterns/strategy.md

```diff
@@ -0,0 +1,153 @@
+or just modify the existing format. 
+So, in order to make chages to our existing code base as much as possible 
+while keeping our solution adoptable to such changes, 
+we might exploit Strategy design pattern. 
+Rust has powerful trait system which allows to implement this pattern. 
+
+## Example
+
+In this example our invariants (or abstractions) are `Context`, `Formatter`, and `Report`. 
+`Text` and `Json` are our strategy structs. 
+These strategies have to implement `Formatter` trait.
+
+```rust
+use std::fmt::Write as FmtWrite;
+use std::{error, result};
+type Result = result::Result<String, Box<dyn error::Error>>;
+
+struct Context {
+    pub keys: Vec<String>,
+    pub values: Vec<i32>,
+}
+
+trait Formatter {
+    fn run(&self, context: &Context) -> Result;
+}
+
+struct Report;
+
+impl Report {
+    fn generate<T: Formatter>(g: T) -> Result {
+        //perform here backend operations which should not bother caller...
+        //fetch data from database
+        let keys = vec!["one".to_string(), "two".to_string()];
+        let values = vec![1, 2];
+        // generate report
+        g.run(&Context { keys, values })
+    }
+}
+
+struct Text;
+impl Formatter for Text {
+    fn run(&self, context: &Context) -> Result {
+        let mut s = String::new();
+
+        for (key, val) in context.keys.iter().zip(context.values.iter()) {
+            write!(&mut s, "{} {}\n", key, val)?;
+        }
+        Ok(s)
+    }
+}
+
+struct Json;
+impl Formatter for Json {
+    fn run(&self, context: &Context) -> Result {
+        let mut s = String::from("[");
+
+        for (key, val) in context.keys.iter().zip(context.values.iter()) {
+            if s.len() > 1 {
+                write!(&mut s, ",")?;
+            }
+            write!(&mut s, "{{\"{}\":\"{}\"}}", key, val)?;
+        }
+        write!(&mut s, "]")?;
+        Ok(s)
+    }
+}
+
+fn main() {
+    assert_eq!(
+        String::from("one 1\ntwo 2\n"),
+        Report::generate(Text).unwrap()
+    );
+
+    assert_eq!(
+        String::from(r#"[{"one":"1"},{"two":"2"}]"#),
+        Report::generate(Json).unwrap()
+    );
+}
+
+```
+
+
+## Advantages
+
+Separation of concerns. In the previous example, Report does not know anything about specific implementations of `Json` and `Text`, whereas the output implementations does not care about how data is preprocessed, stored, and fetched. 
+The only thing they have to know is context and and a specific trait and method to implement, i.e., `Formatter` and `run`. 
+
+## Disadvantages
+
+For each strategy there must me implemented at least one module, so number of modules increases with number of strategies.
+If there are many strategies to choose from then users have to know how strategies differ from one another.
+
+## Discussion
+
+In the previous example all strategies in a single file.
+Typically in Rust, each strategy should be implemented in a separate module file.
```

Comment:
```suggestion
Ways of providing different strategies includes:

- All in one file (as shown in this example)
- Separated as modules, E.g. `formatter::json` module, `formatter::text` module
- Use compiler feature flags, E.g. `json` feature, `text` feature 
- Separated as crates, E.g. `json` crate, `text` crate
```

We may need to link each individual method, especially the use compiler feature flags.